### ISSUE TITLE
Replace LUT/enums with predefined structures

### ISSUE THREAD

**Issue by kammce** [kammce/SJSU-Dev2#396]
Currently we have something like this:

```
/// SSP register lookup table
inline static LPC_SSP_TypeDef * ssp_registers[]  = { LPC_SSP0, LPC_SSP1, LPC_SSP2 };
/// SSP peripherals
enum class Peripheral
{
  kSsp0 = 0,
  kSsp1 = 1,
  kSsp2 = 2,
  kNumberOfPeripherals
};
static constexpr uint8_t kPinSelect[] = { 0b010, 0b010, 0b100 };
static constexpr size_t kNumberOfPeripherals = util::Value(Peripheral::kNumberOfPeripherals);
static constexpr Pin kSspPinMatrix[kNumberOfPeripherals][3] = {
  // SSP0
  { [MatrixLookup::kMosi] = Pin::CreatePin<0, 18>(),
    [MatrixLookup::kMiso] = Pin::CreatePin<0, 17>(),
    [MatrixLookup::kSck]  = Pin::CreatePin<0, 15>() },
  // SSP1
  { [MatrixLookup::kMosi] = Pin::CreatePin<0, 9>(),
    [MatrixLookup::kMiso] = Pin::CreatePin<0, 8>(),
    [MatrixLookup::kSck]  = Pin::CreatePin<0, 7>() },
  // SSP2
  { [MatrixLookup::kMosi] = Pin::CreatePin<1, 1>(),
    [MatrixLookup::kMiso] = Pin::CreatePin<1, 4>(),
    [MatrixLookup::kSck]  = Pin::CreatePin<1, 0>() }
};
static constexpr Lpc40xxSystemController::PeripheralID kPowerBit[] = {
  Lpc40xxSystemController::Peripherals::kSsp0,
  Lpc40xxSystemController::Peripherals::kSsp1,
  Lpc40xxSystemController::Peripherals::kSsp2,
};
```

And this works well, but if you are only using SSP0 then you are taking up space with SSP1 and SSP2's placement in the lookup table. Now this is pretty minor in that it only takes up 8 bytes of additional space, but this happens across the whole L1 directory, so these lookup tables add up. 

An alternative could be something like this:
```
struct Peripheral
{
  struct Ssp0
  {
    LPC_SSP_TypeDef * ssp_registers = LPC_SSP0;
    uint8_t pin_function_select     = 0b010;
    Lpc40xxSystemController::PeripheralID power_bit =
        Lpc40xxSystemController::Peripherals::kSsp0;
    Pin pin[3] = { [MatrixLookup::kMosi] = Pin::CreatePin<0, 18>(),
                   [MatrixLookup::kMiso] = Pin::CreatePin<0, 17>(),
                   [MatrixLookup::kSck]  = Pin::CreatePin<0, 15>() };
  };
  struct Ssp1
  {
    // Same as above but with ssp 1 registers and pins etc...
  }
};

// Using it in actual code
Ssp ssp0(Ssp::Peripheral::Ssp0);
```
With this, space is only allocated if you attempt to pass or store these structures directly. No need to store them in a static lookup table. Using more structures doesn't take up more memory thus there is no dead memory in the code when you attempt to include and use SSP. 

In this case, you only allocate and use what you want and do not get any additional bloat from code/information you are not using. 

EDIT: A better way of doing this is making a single struct object and defining static predefined constexpr objects, like so:

```
  struct Port_t
  {
     uint32_t * reg;
     uint8_t codes;
  };

  // Namespace of available ports
  struct Port
  {
    static constexpr Port_t k1 = 
    {
        .reg = &a,
        .codes = 0b1001,
    };
    static constexpr Port_t k2 =
    {
        .reg = &b,
        .codes = 0b1011,
    };
    static constexpr Port_t k3 =
    {
        .reg = &c,
        .codes = 0b1001,
    };
    static constexpr Port_t k4 =
    {
        .reg = &d,
        .codes = 0b1111,
    };
  };
```

**Comment by kammce** [comment]
Wrote this using godbolt to see the difference between using LUT vs using constexpr static predefined objects:

https://godbolt.org/z/BUAIo2

**Comment by kammce** [comment]
Resolved as of #425
